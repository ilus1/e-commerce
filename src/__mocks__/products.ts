export const mockProducts = [
  { id: '1', name: 'Produto 1', price: 10, photo: 'url1', description: 'Descrição 1' },
];

export const mockCartItems = [
  { id: '1', name: 'Produto 1', price: 10, photo: 'url1', description: 'Descrição 1', count: 1 },
];
