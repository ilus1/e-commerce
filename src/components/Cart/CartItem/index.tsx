import { motion } from 'framer-motion';
import QuantitySelector from '../../QuantitySelector'
import {
  Container,
  CloseButton,
  Price,
  Name,
  Image,
} from './styles';

interface CartItem {
  id: string
  name: string
  price: number
  photo: string
  description: string
  count: number
}

interface CartItemProps {
  item: CartItem
  onRemove: (item: CartItem) => void
  onAddItem: (item: CartItem) => void
  onRemoveItem: (item: CartItem) => void
}

function CartItem({ item, onRemove, onAddItem, onRemoveItem }: CartItemProps) {
  const { photo, name, price } = item;
  return (
    <Container
      data-testid="cart-item-component"
      as={motion.li}
      initial={{ x: 40, opacity: 0 }}
      animate={{ x: 0, opacity: 1 }}
      transition={{
        delay: 0.15,
        duration: 0.8,
        ease: "easeOut"
      }}
    >
      <CloseButton onClick={() => onRemove(item)}>X</CloseButton>
      <Image src={photo} />
      <Name>{name}</Name>
      <QuantitySelector
        item={item}
        onAdd={onAddItem}
        onRemove={onRemoveItem}
      />
      <Price>R$ {price}</Price>
    </Container>
  );
}

export default CartItem;
