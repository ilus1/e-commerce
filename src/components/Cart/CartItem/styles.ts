import styled from "styled-components"

import colors from "../../../constants/colors"

export const Container = styled.li`
  display: flex;
  justify-content: space-between;
  align-content: center;
  align-items: center;
  position: relative;

  min-height: 95px;
  width: 379px;
  border-radius: 8px;
  padding: 0px 20px;
  background-color: ${colors.background};
  box-shadow: -2px 2px 10px 0px rgba(0, 0, 0, 0.05);
`

export const CloseButton = styled.button`
  display: inline-flex;
  align-items: center;
  justify-content: center;
  position: absolute;
  top: -30px;
  right: -5px;

  font-size: 14px;
  font-weight: 400;

  width: 18px;
  height: 18px;
  margin-top: 25px;
  background-color: ${colors.sidebarButtonBackground};
  color: ${colors.background};

  border: none;
  border-radius: 50%;
  cursor: pointer;
`

export const Name = styled.span`
  font-size: 16px;
  font-weight: 400;
  max-height: 36px;
  width: 124px;
  color: ${colors.productDescription};
`

export const Image = styled.img`
  height: 65px;
  object-fit: contain;
`

export const Price = styled.span`
  display: inline-flex;
  font-size: 14px;
  font-weight: 700;
  justify-content: center;
  align-items: center;

  border-radius: 5px;
  padding: 4px 8px;
  color: ${colors.sidebarButtonBackground};
`
