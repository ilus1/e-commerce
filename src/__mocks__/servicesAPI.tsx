import { rest } from "msw";
import { setupServer } from "msw/node";

import { mockProducts } from "./products";

const server = setupServer(
  rest.get('*', (_, res, ctx) => res(ctx.status(200), ctx.json(mockProducts)))
);

export default server;
