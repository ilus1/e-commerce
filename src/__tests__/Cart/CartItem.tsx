import renderer from 'react-test-renderer';
import { render, screen, fireEvent } from '@testing-library/react';
import CartItem from '../../components/Cart/CartItem';
import { mockCartItems } from '../../__mocks__/products';

describe('Componente de um item do carrinho', () => {
  it('Testa carregamento correto do componente', () => {
    const tree = renderer
      .create(
        <CartItem
          item={mockCartItems[0]}
          onRemove={() => {}}
          onAddItem={() => {}}
          onRemoveItem={() => {}}
        />
      );

    expect(tree).toMatchSnapshot();
  });

  it('Testa a função de remoção do item quando o botão de fechar é clicado', () => {
    const mockOnRemove = jest.fn();
    render(
      <CartItem
        item={mockCartItems[0]}
        onRemove={mockOnRemove}
        onAddItem={() => {}}
        onRemoveItem={() => {}}
      />
    );

    fireEvent.click(screen.getByText('X'));
    expect(mockOnRemove).toHaveBeenCalledWith(mockCartItems[0]);
  });
});
