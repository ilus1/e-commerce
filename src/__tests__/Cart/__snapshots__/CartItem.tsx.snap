// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Componente de um item do carrinho Testa carregamento correto do componente 1`] = `
<li
  className="sc-hKgKIp fFLrTl"
  data-testid="cart-item-component"
  style={
    {
      "opacity": 0,
      "transform": "translateX(40px) translateZ(0)",
    }
  }
>
  <button
    className="sc-eCstZk bWQkgF"
    onClick={[Function]}
  >
    X
  </button>
  <img
    className="sc-gKsecS hntZsr"
    src="url1"
  />
  <span
    className="sc-jSguLX dDFcMB"
  >
    Produto 1
  </span>
  <div
    className="sc-bdfCDU fhrPBJ"
  >
    <button
      className="sc-dlfmHC jhxSjB"
      onClick={[Function]}
    >
      -
    </button>
    <p
      className="sc-gsTDqH iORjDB"
    >
      1
    </p>
    <button
      className="sc-dlfmHC jhxSjB"
      onClick={[Function]}
    >
      +
    </button>
  </div>
  <span
    className="sc-iBPTVF jogATO"
  >
    R$ 
    10
  </span>
</li>
`;
